### PR TITLE
SSI Duplicate

### PR DESCRIPTION
**Pull request by n17h31sm**
This PR fixes OWASP/wstg#593.

- [x] This PR handles the issue and requires no additional PRs.

**What did this PR accomplish?**

- Removed the unneeded information from the introduction and merged the sufficient parts into one
- Cleaned the methodology testing paragraphs
- Added some more references

### CODE REVIEW

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 1]
```diff
@@ -6,52 +6,43 @@
...
+Server-Side Includes are directives that the web server parses before serving the page to the user. They represent an alternative to writing CGI programs or embedding code using server-side scripting languages, when there's only need to perform very simple tasks. Common SSI implementations provide directives(commands) to include external files, to set and print web server CGI environment variables, and to execute external CGI scripts or system commands.
 
-Putting an SSI directive into a static HTML document is as easy as writing a piece of code like the following:
+When testing for SSI we are injecting SSI directives as user input and if SSI are enabled and there is no user input validation the server will execute the directive. This is very similar to a classical scripting language injection vulnerability, it occurs when user input is not properly validated and sanitized.
 
-`<!--#echo var="DATE_LOCAL" -->`
+SSI can lead to a Remote Command Execution (RCE), however most webservers have the `exec` directive disabled by default.
 
-to print out the current time.
+### Testing examples
```
Comment (ThunderSon) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:17]: Can we remove this section from this part and use what's mainly needed in the How to Test section?

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 2]
```diff
@@ -6,52 +6,43 @@
...
-
-As in every bad input validation situation, problems arise when the user of a web application is allowed to provide data that makes the application or the web server behave in an unforeseen manner. With regard to SSI injection, the attacker could provide input that, if inserted by the application (or maybe directly by the server) into a dynamically generated page, would be parsed as one or more SSI directives.
-
-This is a vulnerability very similar to a classical scripting language injection vulnerability. One mitigation is that the web server needs to be configured to allow SSI. On the other hand, SSI injection vulnerabilities are often simpler to exploit, since SSI directives are easy to understand and, at the same time, quite powerful, e.g., they can output the content of files and execute system commands.
-
-## Test Objectives
-
-- Identify SSI injection points.
-- Assess the severity of the injection.
+There are more directives and variables which you can find [here](http://www.cheat-sheets.org/sites/ssi.su/)
```
Comment (ThunderSon) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:31]: Refer to this instead: https://httpd.apache.org/docs/current/howto/ssi.html

**Review by ThunderSon — CHANGES_REQUESTED** [review]
I don't feel like the gray testing paragraph is worth it.
Can you remove it and my salvage what could be salvaged, and remove the headers "Black box testing" and "Grey box testing". They're not needed.

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 3]
```diff
@@ -6,52 +6,43 @@
...
-to include the output of a system command.
+Returns the output of the supplied command.
 
-Then, if the web server's SSI support is enabled, the server will parse these directives. In the default configuration, usually, most web servers don't allow the use of the `exec` directive to execute system commands.
-
-As in every bad input validation situation, problems arise when the user of a web application is allowed to provide data that makes the application or the web server behave in an unforeseen manner. With regard to SSI injection, the attacker could provide input that, if inserted by the application (or maybe directly by the server) into a dynamically generated page, would be parsed as one or more SSI directives.
-
-This is a vulnerability very similar to a classical scripting language injection vulnerability. One mitigation is that the web server needs to be configured to allow SSI. On the other hand, SSI injection vulnerabilities are often simpler to exploit, since SSI directives are easy to understand and, at the same time, quite powerful, e.g., they can output the content of files and execute system commands.
-
-## Test Objectives
```
Comment (ThunderSon) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:37]: This is needed. Please set this back in as well :)

**Review by ThunderSon — CHANGES_REQUESTED** [review]

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md (hunk 3 above)
Comment (n17h31sm) [reply document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:37]: This is now moved to the introduction before the how to test

Comment (ThunderSon) [reply document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:37]: This is needed as a section before the ## How to Test. It's the format we parse our checklist against.

Comment (n17h31sm) [reply document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md]: ## Test Objectives section returned

**Comment by ThunderSon** [comment]
Looks good. I will send a commit when you finish your reviews to clean up some minimal stuff (links, refs, etc.)

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 4]
```diff
@@ -41,48 +21,50 @@ This is a vulnerability very similar to a classical scripting language injection
...
+The next step is determining all the possible user input vectors and testing if the SSI Injection is exploitable.
 
-The testing activity required to do this is exactly the same used to test for other code injection vulnerabilities. In particular, we need to find every page where the user is allowed to submit some kind of input, and verify whether the application is correctly validating the submitted input. If sanitization is insufficient, we need to test if we can provide data that is going to be displayed unmodified (for example, in an error message or forum post). Besides common user-supplied data, input vectors that should always be considered are HTTP request headers and cookies content, since they can be easily forged.
+To test this we have to find all the pages where user input is allowed. Possible input vectors may also include Headers and Cookies. Determine how the input is stored and used, i.e if the input is returned as an error message or page element and if it was modified in some way. If we have access to the source code we can easily determine where the input vectors are and how input is handled.
```
Comment (victoriadrake) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:32]:
```suggestion
First find all the pages where user input is allowed. Possible input vectors may also include headers and cookies. Determine how the input is stored and used, i.e if the input is returned as an error message or page element and if it was modified in some way. Access to the source code can help you to more easily determine where the input vectors are and how input is handled.
```

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 5]
```diff
@@ -41,48 +21,50 @@ This is a vulnerability very similar to a classical scripting language injection
...
 
-Once we have a list of potential injection points, we can check if the input is correctly validated and then find out where the provided input is stored. We need to make sure that we can inject characters used in SSI directives:
+Once we have a list of potential injection points, we can check if the input is correctly validated. We need to make sure that we can inject characters used in SSI directives:
```
Comment (victoriadrake) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:34]:
```suggestion
Once you have a list of potential injection points, you may determine if the input is correctly validated. Ensure it is possible to inject characters used in SSI directives:
```

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 6]
```diff
@@ -41,48 +21,50 @@ This is a vulnerability very similar to a classical scripting language injection
...
 
 `< ! # = / . " - > and [a-zA-Z0-9]`
 
-To test if validation is insufficient, we can input, for example, a string like the following in an input form:
+Testing if the SSI Injection is exploitable is very simillar to testing for XSS using
```
Comment (victoriadrake) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:38]:
```suggestion
Testing to see if SSI injection is exploitable is very similar to testing for XSS. You may use a string like the following:
```

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 7]
```diff
@@ -41,48 +21,50 @@ This is a vulnerability very similar to a classical scripting language injection
...
 
-`<!--#include virtual="/etc/passwd" -->`
+`<script>alert("XSS")</script>`
 
-This is similar to testing for XSS vulnerabilities using
+For SSI Injection we can use the directives listed below.
```
Comment (victoriadrake) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:42]:
```suggestion
For SSI injection we can use the directives listed below.
```

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 8]
```diff
@@ -41,48 +21,50 @@ This is a vulnerability very similar to a classical scripting language injection
...
 
-`<script>alert("XSS")</script>`
+`<!--#echo var="VAR" -->`
+
+Returns the value of the variable. [List of SSI include variables](https://httpd.apache.org/docs/current/howto/ssi.html)
```
Comment (victoriadrake) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:46]:
```suggestion
This example returns the value of the variable. [See list of SSI include variables](https://httpd.apache.org/docs/current/howto/ssi.html).
```

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 9]
```diff
@@ -41,48 +21,50 @@ This is a vulnerability very similar to a classical scripting language injection
...
+
+`<!--#include virtual="FILENAME" -->`
+
+If the supplied file is a CGI script the directive will include the output of the CGI script. But the directive may also be used to include the content of a file or list files in a directory.
```
Comment (victoriadrake) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:50]:
```suggestion
If the supplied file is a CGI script, this directive will include the output of the CGI script. This directive may also be used to include the content of a file or list files in a directory.
```

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 10]
```diff
@@ -41,48 +21,50 @@ This is a vulnerability very similar to a classical scripting language injection
...
+
+`<!--#exec cmd="OS_COMMAND" -->`
+
+Returns the output of the supplied command.
+
+There are more directives and variables which you can find [here](https://httpd.apache.org/docs/current/howto/ssi.html)
```
Comment (victoriadrake) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:56]:
```suggestion
You can find more directives and variables [here](https://httpd.apache.org/docs/current/howto/ssi.html).
```

**Review by victoriadrake — COMMENTED** [review]
Copy editing.

**Review by victoriadrake — APPROVED** [review]

**Comment by ThunderSon** [comment]
I did various updates as I read through the document and updated some references.
@n17h31sm please let me know if the changes are good with you as well.

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md [hunk 11]
```diff
@@ -25,62 +25,56 @@ To test for exploitable SSI, inject SSI directives as user input. If SSI are ena
...
 
 The next step is determining all the possible user input vectors and testing to see if the SSI injection is exploitable.
 
 First find all the pages where user input is allowed. Possible input vectors may also include headers and cookies. Determine how the input is stored and used, i.e if the input is returned as an error message or page element and if it was modified in some way. Access to the source code can help you to more easily determine where the input vectors are and how input is handled.
 
-Once you have a list of potential injection points, you may determine if the input is correctly validated. Ensure it is possible to inject characters used in SSI directives:
+Once you have a list of potential injection points, you may determine if the input is correctly validated. Ensure it is possible to inject characters used in SSI directives such as `<!#=/."->` and `[a-zA-Z0-9]`
 
-`< ! # = / . " - > and [a-zA-Z0-9]`
+The below example returns the value of the variable. Check the [references](#references) section to better assess the system at hand with every server's documentation.
```
Comment (victoriadrake) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md:36]:
```suggestion
The below example returns the value of the variable. The [references](#references) section has helpful links with server-specific documentation to help you better assess a particular system.
```

**Comment by n17h31sm** [comment]
I checked the changes, it seems fine to me as well. Maybe just also commit @victoriadrake proposed change.

**Review by kingthorin — APPROVED** [review]
Thanks

**Review by ThunderSon — APPROVED** [review]
Vic's way with words is always better, no doubts about that!

**Comment by victoriadrake** [comment]
Glad to be able to help! Thank you for your contribution, @n17h31sm! Congratulations on the merge! :tada:

**Pull request merged**